### PR TITLE
perf(filecoin-proofs): speed up Fr32Reader

### PR DESCRIPTION
**Pull request by dignifiedquire**
Switch to processing bit padding in blocks of 127bytes using u128s.

This also improves correctness, as now the reader will emit a full `Fr32` at the end, not just a single byte with the additional bits, which is why some of the test code was updated.

Speed differences can be seen here, with the diffs being the same benchmarks run against master. https://gist.github.com/dignifiedquire/1c21680f237a8f3aa5ad43f0008ffa41

### CODE REVIEW

**Review by cryptonemo — APPROVED** [review]
I reviewed this earlier and it looks good, thanks for rebasing.

File: filecoin-proofs/src/fr32_reader.rs [hunk 1]
```diff
@@ -160,146 +108,44 @@ impl<R: io::Read> io::Read for Fr32Reader<R> {
...
-    }
+        let mut read = 0;
 
-    #[cfg(target_endian = "little")]
-    pub fn read_u16_into(&mut self, target: &mut [u8]) {
-        assert!(target.len() >= 2);
+        while read < num_bytes {
+            // load new block
+            if self.to_process == 0 {
+                let bytes_read = self.fill_buffer()?;
```
Comment (porcuquine) [inline filecoin-proofs/src/fr32_reader.rs:118]: This fills `in_buffer`.

File: filecoin-proofs/src/fr32_reader.rs [hunk 2]
```diff
@@ -1,157 +1,105 @@
...
-#[derive(Debug)]
 pub struct Fr32Reader<R> {
     /// The source being padded.
     source: R,
-    /// How much of the target already was `read` from, in bits.
-    target_offset: u64,
-    /// Currently read byte.
-    buffer: Buffer,
+    /// Currently read block.
+    in_buffer: [u8; NUM_U128S_PER_BLOCK * 16],
```
Comment (porcuquine) [inline filecoin-proofs/src/fr32_reader.rs:8]: Instead of 16, you could explicitly calculate with `size_of` to show that the in and out buffers hold the same number of bytes.

File: filecoin-proofs/src/fr32_reader.rs [hunk 3]
```diff
@@ -160,146 +108,44 @@ impl<R: io::Read> io::Read for Fr32Reader<R> {
...
+            let out_end = out_start + len;
 
-        self.pos += 64;
-        self.data
-    }
+            target[start..end]
+                .copy_from_slice(&self.out_buffer.as_byte_slice()[out_start..out_end]);
+            read += len;
+            self.out_offset += len;
+            self.to_process -= div_ceil(len * 8, 256);
```
Comment (porcuquine) [inline filecoin-proofs/src/fr32_reader.rs:145]: `div_ceil(len, 32)` might be clearer. But this raises a question, if `to_process` is the number of blocks, and one block is 4 `Fr`, shouldn't this be `div_ceil(len, 128)`?

I can believe the actual algorithm is doing the right thing here, but I'm confused trying to line up all the names and descriptions throughout this whole changeset.

I think maybe 'block' in this context is being used to mean 'one `Fr`s worth of bytes' — but it seems like in the definition of `NUM_U128S_PER_BLOCK` (where the value is 8), 'block' is being used to refer to 128 bytes = 4 `Fr32`s.

Am I missing something, or can some rectification of names help?

**Review by porcuquine — COMMENTED** [review]
This generally looks good, and I can't see any definite problems. However, the naming throughout is confusing enough that I find it hard to follow the logic. In some cases, the naming actively confuses me because words _seem_ to be used ambiguously. If you can clarify such cases by using less ambiguous naming and especially avoid cases where the same word is being used to mean entirely different things (if any such actually exist), that would be valuable. Although it seems like a small thing, once such naming gets locked in, it can be very expensive for future readers of the code to swap in the context required to understand it.

File: filecoin-proofs/src/fr32_reader.rs [hunk 4]
```diff
@@ -1,157 +1,105 @@
...
+    to_process: usize,
+    /// How many bytes of the out_buffer are already read out.
+    out_offset: usize,
     /// Are we done reading?
     done: bool,
 }
 
+const NUM_U128S_PER_BLOCK: usize = 8;
+const MASK_SKIP_HIGH_2: u128 = 0b0011_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111;
+
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply filecoin-proofs/src/fr32_reader.rs:21]: I find that much harder to inspect to be honest

File: filecoin-proofs/src/fr32_reader.rs [hunk 5]
```diff
@@ -1,157 +1,105 @@
...
-    target_offset: u64,
-    /// Currently read byte.
-    buffer: Buffer,
+    /// Currently read block.
+    in_buffer: [u8; NUM_U128S_PER_BLOCK * 16],
+    /// Currently writing out block.
+    out_buffer: [u128; NUM_U128S_PER_BLOCK],
+    /// How many blocks are left in the buffers.
+    to_process: usize,
+    /// How many bytes of the out_buffer are already read out.
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply filecoin-proofs/src/fr32_reader.rs:13]: Well the method is called `read`, so for me data is getting "read out" of this buffer, by writing it into the `target`.

File: filecoin-proofs/src/fr32_reader.rs [hunk 6]
```diff
@@ -397,9 +185,10 @@ mod tests {
         assert_eq!(&padded[0..31], &data[0..31]);
         assert_eq!(padded[31], 0b0011_1111);
         assert_eq!(padded[32], 0b0000_0011);
-        assert_eq!(padded.len(), 33);
-
-        assert_eq!(padded.into_boxed_slice(), bit_vec_padding(data));
+        assert_eq!(padded.len(), 64);
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply filecoin-proofs/src/fr32_reader.rs:188]: As mentioned in the PR description, there is a correctness fix in this PR, before the Reader would produce trailing bits by a single byte, now it always produces a full Fr32, which is why the total padded length increases to the next multiple of 32 bytes.

File: filecoin-proofs/src/fr32_reader.rs [hunk 7]
```diff
@@ -160,146 +108,44 @@ impl<R: io::Read> io::Read for Fr32Reader<R> {
...
-    #[cfg(target_endian = "little")]
-    pub fn read_u32_into(&mut self, target: &mut [u8]) {
-        assert!(target.len() >= 4);
-        let value = self.read_u32().to_le_bytes();
-        target[0] = value[0];
-        target[1] = value[1];
-        target[2] = value[2];
-        target[3] = value[3];
-    }
+            let start = read;
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply filecoin-proofs/src/fr32_reader.rs:134]: Well it is very unfortunate that the past tense and the verb happen to be the same word here, as this is explicitly the past tense, maybe "bytes_read" would be better?

File: filecoin-proofs/src/fr32_reader.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (porcuquine) [reply filecoin-proofs/src/fr32_reader.rs:21]: Really? It's easy for you to see at a glance that this is wrong?

```
0b0011_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111
```

File: filecoin-proofs/src/fr32_reader.rs [hunk 8]
```diff
@@ -160,146 +108,44 @@ impl<R: io::Read> io::Read for Fr32Reader<R> {
...
+                if bytes_read == 0 {
+                    self.done = true;
+                    break;
+                }
 
-    #[cfg(target_endian = "little")]
-    pub fn read_u32(&mut self) -> u32 {
-        debug_assert!(self.available() >= 32);
+                self.process_block();
+                self.to_process += div_ceil(bytes_read * 8, 254);
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply filecoin-proofs/src/fr32_reader.rs:127]: To process is in number of blocks. `bytes_read` is in `bytes`, so we multiply it by `8` to get bits. Because this is an input block its size is `254`, so we divide it by that, which is why I do not want to divide by `32`, even though this might be the same thing in the cases this code covers.

File: filecoin-proofs/src/fr32_reader.rs (hunk 3 above)
Comment (dignifiedquire) [reply filecoin-proofs/src/fr32_reader.rs:145]: to process is in number of Frs sorry, which is 254 bits in, 256bits out

File: filecoin-proofs/src/fr32_reader.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply filecoin-proofs/src/fr32_reader.rs:21]: neither of them is "easy" for me, but in thase case I simply count the 1s and know if it is good, but your notation requires me to think through bitshfiting rules, and how the subtraction operator is used on a bit level.

[the remark this answers is not in the document]
Comment (porcuquine) [reply filecoin-proofs/src/fr32_reader.rs:50]: My main point was to avoid counting the `1`s — not the specific construction used to avoid it. Especially with an appropriate comment, I'm not averse to asking readers to think through bitshifting rules.

A simpler version which requires less thinking and still no counting would be:

```
const MASK_SKIP_HIGH_2: u128 =  u128::MAX >> 2;
```

Maybe that is the best of both worlds?
